core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala

```diff
@@ -751,6 +751,14 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
    * Visible for testing
    */
   private[history] def doMergeApplicationListing(
+      reader: EventLogFileReader,
+      scanTime: Long,
+      enableOptimizations: Boolean,
+      lastEvaluatedForCompaction: Option[Long]): Unit = doMergeApplicationListingInternal(
+    reader, scanTime, enableOptimizations, lastEvaluatedForCompaction)
+
+  @scala.annotation.tailrec
+  private def doMergeApplicationListingInternal(
```

Comment:
Otherwise:

```
[error] /.../spark/core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:754:24: could not optimize @tailrec annotated method doMergeApplicationListing: it is neither private nor final so can be overridden
[error]   private[history] def doMergeApplicationListing(
[error]                        ^
[error] one error found
```